### ISSUE TITLE
Tab complete for class __call__ not working

### ISSUE THREAD

**Issue by MarcoGorelli** [davidhalter/jedi#1962]
What follows is an issue I originally reported to IPython, but was asked to report here (original issue: https://github.com/ipython/ipython/issues/14184)

---

If I make the following two files:
```python
# t.py
class MyClass:
    def my_method(self) -> None:
        print('this is my method')


def foo() -> MyClass: return MyClass()

```
and
```python
# f.py
class MyClass:
    def my_method(self) -> None:
        print('this is my method')

class Foo:
    def __call__(self) -> MyClass:
        return MyClass()

foo = Foo()
```

then `foo().my_method()` works for both

But - I only get the auto-complete suggestion for the first one: `ipython -i t.py`:

![Screenshot 2023-09-30 102929](https://github.com/ipython/ipython/assets/33491632/9d2cc19b-f899-410f-bbd5-4333b0ce5ad5)

`ipython -i f.py`:

![Screenshot 2023-09-30 102832](https://github.com/ipython/ipython/assets/33491632/d29cadab-b4c8-4bc4-9dfb-fe7cf0f8e0d0)

Shouldn't both autocomplete to the method available in `MyClass`?

For reference, this came up here https://github.com/pola-rs/polars/issues/11433

**Comment by davidhalter** [comment]
I agree that this should probably work, but it's not a normal Jedi case due to the fact that the source code is not available anymore (for Jedi). I feel like we should fix this, but at the same time I understand very well why this didn't work in the first place.
